src/Microsoft.OData.Core/PropertyCache.cs

```diff
@@ -14,15 +15,68 @@ namespace Microsoft.OData
     /// </summary>
     internal class PropertyCache
     {
-        private readonly Dictionary<string, PropertySerializationInfo> propertyInfoDictionary = new Dictionary<string, PropertySerializationInfo>();
+        private readonly Dictionary<PropertyKey, PropertySerializationInfo> propertyInfoDictionary = new Dictionary<PropertyKey, PropertySerializationInfo>();
 
-        public PropertySerializationInfo GetProperty(IEdmModel model, string name, string uniqueName, IEdmStructuredType owningType)
+        private struct PropertyKey : IEquatable<PropertyKey>
+        {
+            public PropertyKey(int depth, string name, string fullTypeName)
+            {
+                this.Depth = depth;
+                this.Name = name;
+                this.FullTypeName = fullTypeName;
+            }
+
+            public int Depth { get; private set; }
+
+            public string Name { get; private set; }
+         
+            public string FullTypeName { get; private set; }
+
+            public static bool operator ==(PropertyKey lhs, PropertyKey rhs)
+            {
+                return lhs.Equals(rhs);
+            }
+
+            public static bool operator !=(PropertyKey lhs, PropertyKey rhs)
+            {
+                return !lhs.Equals(rhs);
+            }
+
+            public override int GetHashCode()
+            {
+                return this.Depth.GetHashCode() ^

```

Comment:
Not an expert but saw comments that XOR is not the best mixing function. 
Newer SDKs provide System.Hash.Combine methods that use multiplication and addition.

https://source.dot.net/#System.Reflection.Metadata/System/Reflection/Internal/Utilities/Hash.cs,10
https://source.dot.net/#System.Private.CoreLib/HashCode.cs,76
https://docs.microsoft.com/en-us/dotnet/api/system.hashcode.add?view=net-5.0

Reply:
Unfortunately we cannot use that API, need a common solution for all target runtimes. Honestly I have used this hash code pattern and never found any issues. However I would not be surprised if this is not efficient for very large collections, where collisions become more likely.

Reply:
If there is a better hash code implementation that is common to all runtimes, let me know and will change.